### PR TITLE
Document Jekyll's Philosophy

### LINKED ISSUE #5342: Jekyll's Philosophy
Hey y'all! I wanted to put together a guiding philosophy for the Jekyll project to help make it easier to decide whether to accept a feature request, whether a bug is a bug or actually a feature request, and so on. Jekyll has a massive user base (arguably the largest) and we need to do our best to serve the needs of our users. We have some philosophy verbiage in [the README](https://github.com/jekyll/jekyll#philosophy) and in [the maintainer documentation](https://github.com/jekyll/jekyll/blob/master/docs/triaging-an-issue.md), but I think we should create a `PHILOSOPHY` document everyone can refer to as we iterate on Jekyll to help guide us. This is meant to be a document that _can_ change, but think of it as the Constitution rather than a handful of malleable laws.

I wanted to open an issue to help facilitate a conversation around one question: what is Jekyll's philosophy? In other words, what are our guiding principles?

To start the conversation, here are a few I keep in mind as I look at issues and evaluate pull requests:
- No magic: Jekyll should do what you ask of it and nothing more.
- Easy out-of-the-box experience: If something increases the learning curve, we should be skeptical. Things should "just work" with sane defaults.
- Jekyll is content-focused: what matters is the content. Let the content be king.
- Jekyll is not an airplane cockpit: take on the burden of making hard choices on our users' behalf. Do not give a configuration option for every little thing.
- Jekyll is for Hackers: Jekyll should be easily extensible.

What do you think of as Jekyll's philosophical guiding principles?

Prior art: [WordPress](https://wordpress.org/about/philosophy/).

/cc @jekyll/build @jekyll/documentation @jekyll/ecosystem @jekyll/performance @jekyll/stability @jekyll/windows @jekyll/core

### PR DESCRIPTION
**Pull request by parkr**
Here are 5 points to start with.

Fixes #5342.

/cc @benbalter @pathawks @mattr- for :eyes: especially. :heart:

### CODE REVIEW

**Comment by ashmaroli** [comment]
Looks Good! But hows is this document *linked* to rest of the documentation at `jekyllrb.com`
I understand that this pertains to Jekyll Core, but how much of this is applicable to **Jekyll Plugins** created by the contributors?

File: docs/philosophy.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+Jekyll offers a unique philosophy when approaching the problem of static
+site generation. This core philosophy drives development and product
+decisions. When a contributor, maintainer, or user asks herself what Jekyll
+is about, the following principles should come to mind:
+
+### 1. No Magic
+
+Jekyll is not magic. A user should be able to understand the underlying
+processes that make up the Jekyll build without much reading. It should
+behave "as you'd expect."
```
Comment (benbalter) [inline docs/philosophy.md]: @parkr you've described this previously as "Jekyll does exactly what the user asks, no more, no less". Is that still accurate? "I expect this" is very different from "I asked you to do this" or "I understand what's going on under the hood to cause this to happen".

File: docs/philosophy.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+### 2. It "Just Works"
+
+The out-of-the-box experience should be that it "just works." Run `gem
+install jekyll` and it should build any Jekyll site that it's given.
+Features like auto-regeneration and settings like the markdown renderer
+should represent sane defaults that work perfectly for the vast majority of
+cases. The burden of configuration should not be placed on the user.
```
Comment (benbalter) [inline docs/philosophy.md]: > The burden of configuration

Initial configuration? Initial setup? Configuration in general?

File: docs/philosophy.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+Why is Jekyll so loved by content creators? It focuses on content first and
+foremost, making the process of publishing content on the Web easy. Users
+should find the management of their content enjoyable and simple.
+
+### 4. Stability
+
+If a user's site builds today, it should build tomorrow.
+Backwards-compatibility should be strongly preferred over breaking changes.
+Upon breaking changes, provide a clear path for users to upgrade.
```
Comment (benbalter) [inline docs/philosophy.md]: Also maybe that when it comes to breaking changes, practicality over purity. There should be a strong reason to institution a breaking change, and it needs to outweigh the burden we're pushing on users. "I would have done it this way", "this way is easier to read", or "this is the textbook approach" aren't compelling reasons to break a user's site.

File: docs/philosophy.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+### 5. Small & Extensible
+
+The core of Jekyll should be simple and small, and extensibility should be
+a first-class feature to provide added functionality from community
+contributors. The core should be kept to features used by at least 90% of
+users–everything else should be provided as a plugin.
```
Comment (benbalter) [inline docs/philosophy.md]: Maybe to note that (A) Jekyll itself maintains plugins, and (B) the preferred approach for new features is to create the necessary APIs first and then dogfood an implementation, rather that going in the reverse direction. Put another way, Jekyll should use its own plugin API system to develop new features, it's not just for third-party functionality.

File: docs/philosophy.md (hunk 1 above)
Comment (parkr) [inline docs/philosophy.md]: >  "I expect this" is very different from "I asked you to do this"

This is a great point! I'll update this piece to clarify that.

File: docs/philosophy.md (hunk 2 above)
Comment (parkr) [inline docs/philosophy.md]: I think **initial configuration** is what I meant here. It's the out-of-the-box experience that I'm getting at.

**Comment by parkr** [comment]
https://jekyllrb.com/philosophy

**Comment by benbalter** [comment]
Nice! Perhaps this warrants a blog post? 😄

**Comment by parkr** [comment]
Good idea!

**Pull request merged**